### PR TITLE
Adding Cross Validation Code, Preprocessing and Weighted Loss 

### PR DESCRIPTION
**Pull request by trenaudie**
## Description 

This PR adds the functionality for: 
- **Cross-Validation**: K - Fold cross validation over the training set, with optional inferences over the validation set at each epoch during training. The final IOU metric is computed over the oof predictions, spanning the entire training set. This maximises the number of OOF points over which this metric is computed and therefore minimizes the variance of the IOU metric estimator for our model. 
- **Unet3D model** : this model is perfectly tailored for our use case : a segmentation task over T time steps, and H * W images. The conv3D kernels also span the 10 input channels. This model outputs a logit for each class, for each of the T timesteps, and pixel locations. We then choose the median of the predictions over the T timesteps . 
- **Preprocessing** : We removed the Clouds from the images by removing any image with too many "outlier" pixels, defined as pixels whose distance to the median pixel is too high. 
- **Weighted Loss**:  We add weights to the cross entropy loss, putting more weight on the parcels with less coverage during training. This may help the model better obtain better IOU scores for these less common classes.

### CODE REVIEW

**Comment by trenaudie** [comment]
Bonsoir Louis,
Excuse-moi, je pensais faire une PR sur ma fork, pour merger ma branch locale sur la branche main de mon repo forké. 
Malheureusement, je viens d'apprendre que la PR s'est ouverte sur ton répo. 
Tu n'es pas obligé de merge, évidemment. 
A demain !

**Comment by LouisStefanuto** [comment]
Pas de soucis ;)
Hâte de voir votre approche !

Le dim. 20 oct. 2024, 19:06, Tanguy RENAUDIE ***@***.***> a
écrit :
